src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs

```diff
@@ -407,5 +406,238 @@ public static void SetLastWriteTime(string fullPath, DateTimeOffset time, bool a
+                    Debug.Assert(Path.IsPathFullyQualified(targetPath));
+                    Debug.Assert(targetPath[1] == '?');
+                    targetPath[1] = '\\';
+                }
+                else if (normalize)
+                {
+                    // Target path is relative and is for ResolveLinkTarget(), we need to append the link directory.
+                    return Path.Join(Path.GetDirectoryName(linkPath.AsSpan()), targetPath);
+                }
+
+                return targetPath.ToString();
+            }
+            finally
+            {
+                if (buffer != null)
+                {
+                    ArrayPool<byte>.Shared.Return(buffer);
+                }
+            }
+        }
+
+        private static unsafe string? GetFinalLinkTarget(string linkPath, bool isDirectory)
+        {
+            Interop.Kernel32.WIN32_FIND_DATA data = default;
+            GetFindData(linkPath, isDirectory, ref data);
+
+            // The file or directory is not a reparse point.
+            if ((data.dwFileAttributes & (uint)FileAttributes.ReparsePoint) == 0 ||
+                // Only symbolic links are supported at the moment.
+                (data.dwReserved0 & Interop.Kernel32.IOReparseOptions.IO_REPARSE_TAG_SYMLINK) == 0)
+            {
+                return null;
+            }
+
+            // We try to open the final file since they asked for the final target.
+            using SafeFileHandle handle = OpenSafeFileHandle(linkPath,
+                    Interop.Kernel32.FileOperations.OPEN_EXISTING |
+                    Interop.Kernel32.FileOperations.FILE_FLAG_BACKUP_SEMANTICS);
+
+            if (handle.IsInvalid)
+            {
+                // If the handle fails with "not found", is becuse the link was broken.
+                // We need to fallback to manually traverse the link targets and return the target of the last resolved link.
+                int error = Marshal.GetLastWin32Error();
+                if (error == Interop.Errors.ERROR_FILE_NOT_FOUND ||
+                    error == Interop.Errors.ERROR_PATH_NOT_FOUND)
+                {
+                    return GetFinalLinkTargetSlow(linkPath);
+                }
+
+                throw Win32Marshal.GetExceptionForWin32Error(error, linkPath);
+            }
+
+            const int InitialBufferSize = 4096;
+            char[]? buffer = ArrayPool<char>.Shared.Rent(InitialBufferSize);
+            try
+            {
+                uint result = GetFinalPathNameByHandle(handle, buffer);
+
+                // If the function fails because lpszFilePath is too small to hold the string plus the terminating null character,
+                // the return value is the required buffer size, in TCHARs. This value includes the size of the terminating null character.
+                if (result > InitialBufferSize)
+                {
+                    ArrayPool<char>.Shared.Return(buffer);
+                    buffer = null;
```

Comment:
I was thinking that if the `ArrayPool<char>.Shared.Rent` in the next line failed for any reason, we would end up with an unreturned array in this scope. I guess that's not a problem since `buffer` is local. I will get rid of it, thanks.